### DISCUSSION TITLE
Question: moving away from sqlite to a more performant DB?

### DISCUSSION THREAD

**Discussion by klntsky** [CardanoSolutions/kupo#147] [Q&A]
### What is your idea? Provide a use case.

In light of #131 which we also observed in production, I really want to ask a question: why is sqlite used as database backend for Kupo?

Do I understand it correctly that the need to rebuild the index after rollbacks is due to sqlite being used? We are considering forking kupo with the goal of replacing sqlite with postgres, what do you think?

[a comment by an automated account is not shown]

**Comment by KtorZ** [comment]
> @klntsky: why is sqlite used as database backend for Kupo?

Because it works, it's embedded and has great performances. So Kupo exists as a single black-box with no dependency. It's simple to install, simple to use and has very decent performances. Also, Kupo has always been meant as a backend service for single-dapps, but not necessarily to be used to power an entire multi-dapps infrastructure. The idea is that it's a suitable choice to embed alongside your dapp backend, but clearly not meant for servicing an API query layer for millions of users. 

> @klntsky: Do I understand it correctly that the need to rebuild the index after rollbacks is due to sqlite being used?

I am not sure where this is coming from? Do you mean, the index as in 'SQLite table indexes' or as in 'Kupo'? If the former, then, it's true that until 2.5.0, Kupo used to create some indexes on-the-fly for rollbacks but it was proven to lead to a lot of unnecessary I/O operations and indexes are now created only once during the initial start-up. 

> @klntsky: considering forking kupo with the goal of replacing sqlite with postgres

Not much interest. One of the design goal of Kupo is to be **simple**. I've flirted with the idea of supporting multi-backends but in my experience, it has never really worked well for projects. It adds complexity without much benefits. If you need something where you want to pick your own backend solution, I'd recommend looking into solutions like [Oura](https://github.com/txpipe/oura). Kupo is opinionated and that's a conscious choice. 

Also, I am not sure why many people hate SQLite this much and see it as a "poor choice" for a backend. In particular, we have done benchmarks against PostgreSQL (cardano-db-sync's model) for a variety of queries and Kupo w/ SQLite was one or two order of magnitudes faster in pretty much all cases.  

One solution that I want to explore, however, is [ducksdb](https://duckdb.org/) for being also a simple and embedded database engine; yet a columnar one. For the type of queries that are performed by Kupo, this could work quite fantastically.

**Reply by theeldermillenial** [reply]
I would be very interested in attending duckdb implementation.

**Comment by biofreeze** [comment]
> The idea is that it's a suitable choice to embed alongside your dapp backend, but clearly not meant for servicing an API query layer for millions of users.

@KtorZ I was thinking of using Fly.io to spin up VM's very quickly. I figure I can download an almost 100% synced Kupo snapshot that I've created, and start servicing queries pretty quickly? I should be able to use Fly.io to horizontally scale Kupo queries? By using snapshots, I should be able to start accepting queries within only a few minutes.

Thoughts? 😃

**Comment by biofreeze** [comment]
Also @KtorZ any thoughts about allowing a remote connection string instead of `workdir`? I've been exploring https://turso.tech which I believe uses SQLite under the hood 👍

**Reply by KtorZ** [reply]
Could be an interesting experiment indeed. Out of curiosity, what are you pursuing with that? Further reducing the local hardware / storage requirements?

**Comment by uhbif19** [comment]
>  I've flirted with the idea of supporting multi-backends but in my experience, it has never really worked well for projects.

Why? SQL has a known standard, and some subset of it is supported by all major DB engines. Do you use some SQLite-specific features?

**Reply by KtorZ** [reply]
All major db engines have their subtleties and differences. If you only take SQLite, MySQL & PostgreSQL; they each have their own types and specificity. Surely, you can find an intersection where they all work, but truth is that a backend trying to do this must either forfeit interesting features of one or the other engine, or must embrace the added complexity of supporting them all. 

Don't get me wrong, it is far easier to support multiple SQL backend that it is to support, say, an SQLite backend and a MongoDB backend. 

Plus, I still haven't seen good evidences that PostgreSQL would perform _that much better_ that it's worth absorbing that complexity. Quite the opposite. In all benchmarks we've done against similar PostgreSQL solutions, SQLite was faster. And more important, it defies one of the core principles of Kupo: simplicity and ease of use.

**Reply by uhbif19** [reply]
> Plus, I still haven't seen good evidences that PostgreSQL would perform that much better that it's worth absorbing that complexity

I am not sure that performance is the only point. For me other reasons are:

1. Many people would like to use managed databases for anything resembling production setup. It is simpler to have it reliable. Even if data is not 
2. Anything remotely resembling distributed application would need a remote connection to DB (AFAIU). For example, if you have one server for real-time tx construction and second for showing DApp stats, you need remote connection. And even if you do not have to do this now you may need to anticipate such scaling.
3. You mentioned `DuckDB`, which is OLAP engine. I think for analytic usecases it is known pattern to use separate replica (of same or OLAP DB), for analytic queries (to not interfere with performance of regular transactions for example).

As for interoperability, I would very much understand your point if it was about MySQL vs PostgreSQL. I am aware of at least two apps hard-coding such choice: Hasura and PostgREST. But I was always under impression, that SQLite is (almost?) always providing subset of their features in terms of SQL queries possible. And AFIAU kupo is not using DB-specific features like specific of communicating protocol or stored procedures, like Hasura does. So I for now do not understand why this point is applicable in kupo case.
